### PR TITLE
feat(python): add "calamine" support to `read_excel`, using `fastexcel` (~8-10x speedup)

### LINKED ISSUE #13874: Add "calamine" and (maybe) "rxls" engines to read_excel
### Description

Hello! 

I've finally published my `rxls` xlsx/xlsb reading engine to [PyPi](https://pypi.org/project/rxls/), after a long time, and extensive testing on real files.

Also, I've update [benchmarks](https://github.com/SaelKimberly/rxls/blob/main/benchmarks.ipynb), and add `calamine` engine to it. Maybe, you will be interested 😃

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #13874.

A long-standing request has been calamine bindings for `read_excel`; the last time I looked at this `python_calamine`[^1] was suggested, but there were some issues, so I didn't go ahead. I took another look at the python/calamine landscape this week and found both that `python_calamine` had _significantly_ improved and would have been suitable for integration, but also discovered that there is a better alternative: `fastexcel`[^2].

Aside from the clean code, this library can skip materialisation to Python and provide us Arrow data, leading to further large speedups over `python_calamine` for our use-case. Originally `fastexcel` supported Python >= 3.10, but I made a [PR](https://github.com/ToucanToco/fastexcel/pull/152) to extend support to earlier versions so that it matches our own version matrix (>= 3.8). 

My thanks to @lukapeschke for reviewing, accepting, and turning out a new `fastexcel` release that contains the extended compatibility within ~24 hours! 🎉

## Note

Currently the `fastexcel` module is available for Linux and macOS; no Windows build yet. (**Update:** but Windows support is coming soon to `0.8.0` - see  https://github.com/ToucanToco/fastexcel/pull/157).

## Example

```python
from codetiming import Timer
import pandas as pd
import polars as pl

# generate five 1,000,000 row columns of various types
df = pl.DataFrame({
    "v": range(1_000_000), 
    "w": 999.999, 
    "x": -42.0, 
    "y": date.today(), 
    "z": "acbdefghijlmnop",
})

# create an excel file to read from
wb = df.write_excel(xl := "test.xlsx")

# benchmark the different engines
with Timer():
    pf0 = pd.read_excel(xl)
with Timer():
    df0 = pl.read_excel(xl, engine="xlsx2csv")
with Timer():
    df1 = pl.read_excel(xl, engine="openpyxl")
with Timer():
    df2 = pl.read_excel(xl, engine="calamine")  # << newly integrated
```
read_excel_bench

_(Hardware: Apple Silicon M2 Max)_

**Result:** Roughly an order of magnitude speedup over our current options, with better than average type inference.

## Also

* Started making the `read_excel` parameters more generic, deprecating "xlsx2csv_options" in favour of "engine_options".
* Reduced the amount of copy/paste code we have when importing optional modules by introducing a new `import_optional` utility function that replaces the current boilerplate with a simple one-liner.

## Short-term follow-ups

* The fastexcel/calamine engine can replace our existing `ods` reader (`ezodf` + `lxml`). I intend to make `fastexcel` the new default engine for `read_ods` as `ezodf` is [unmaintained](https://pypi.org/project/ezodf/#history), old, and _much_ slower.
* It looks like fastexcel/calamine can also support the `.xls` format (the older flavour of Excel file, as compared to `.xlsx`). This should work out of the box, but I can add some magic-bytes detection to automatically pass such files to the newer engine (if none was specified).

## Longer-term

* We may be able to push the speed even further by integrating _directly_ with Calamine down in the lower-levels of the Rust engine (which would also simplify loading Excel data into Polars for our Rust users). 
* Until/unless we decide to do that, we can look to replace the current default engine with fastexcel/calamine once it becomes more featureful (at the moment there are limited features, but hopefully this will change!)

[^1]: python_calamine: https://github.com/dimastbk/python-calamine 
[^2]: fastexcel: https://github.com/ToucanToco/fastexcel

### CODE REVIEW

**Comment by alexander-beedie** [comment]
(I have a few tests to fix up after adding windows skips, but it's 02:00 here, so I'll fix them in the morning 😅)

**Comment by avimallu** [comment]
Holy smokes. This is amazing. I don't use Excel much anymore, but if I had this "back in the day" 😍.

Great work!

**Comment by ritchie46** [comment]
Great stuff! 

> We may be able to push the speed even further by integrating directly with Calamine down in the lower-levels of the Rust engine (which would also simplify loading Excel data into Polars for our Rust users).

I still think we must do this. As I like to have a good architecture for us to support many (more) exotic readers without bloating our binary. We can create `polars-arrow` chunks and accept predicate and projection pushdown.

**Comment by alexander-beedie** [comment]
> I still think we must do this. As I like to have a good architecture for us to support many (more) exotic readers without bloating our binary. We can create `polars-arrow` chunks and accept predicate and projection pushdown.

Yup; and now we have a strong benchmark to aim for! 🤣 Lower-level integration could lead to some really special performance, and we can look to be fast by default once we hot a sufficient set of features.

File: py-polars/polars/io/spreadsheet/functions.py [hunk 1]
```diff
@@ -738,13 +747,49 @@ def _read_spreadsheet_openpyxl(
...
+
+def _read_spreadsheet_calamine(
+    parser: Any,
+    sheet_name: str | None,
+    read_csv_options: dict[str, Any] | None,
+    schema_overrides: SchemaDict | None,
+    *,
+    raise_if_empty: bool,
+) -> pl.DataFrame:
+    ws = parser.load_sheet_by_name(sheet_name)
```
Comment (lukapeschke) [inline py-polars/polars/io/spreadsheet/functions.py:761]: NOTE: Once calamine 0.24.0 is out and https://github.com/ToucanToco/fastexcel/pull/147 is polished and merged, you could use one of the new eager loading methods to lower the memory footprint here

**Review by lukapeschke — COMMENTED** [review]
Looks great :rocket: I'll try to work on windows support ASAP

> I intend to make `fastexcel` the new default engine for `read_ods`

That would be really cool, but our tests on `.ods` files are really minimal for now, so increasing test coverage for that on our side would be needed

**Review by ritchie46 — APPROVED** [review]

**Comment by alexander-beedie** [comment]
> That would be really cool, but our tests on `.ods` files are really minimal for now, so increasing test coverage for that on our side would be needed

Yup; I didn't pull the trigger on that with this PR yet - however, the competition is a module that hasn't been updated [in 8 years](https://pypi.org/project/ezodf/#history) and only had 7 releases. It's a low bar to beat 😅

File: py-polars/polars/io/spreadsheet/functions.py [hunk 2]
```diff
@@ -517,23 +518,32 @@ def _initialise_spreadsheet_parser(
...
-            raise ImportError(msg) from None
+        openpyxl = import_optional("openpyxl")
         parser = openpyxl.load_workbook(source, data_only=True, **engine_options)
         sheets = [{"index": i + 1, "name": ws.title} for i, ws in enumerate(parser)]
         return _read_spreadsheet_openpyxl, parser, sheets
 
+    elif engine == "calamine":
+        # note: can't read directly from bytes (yet) so
+        if read_bytesio := isinstance(source, BytesIO):
+            temp_data = NamedTemporaryFile(delete=True)
```
[the remark this answers is not in the document]
Comment (alexander-beedie) [reply py-polars/polars/io/spreadsheet/functions.py:529]: That would be great :)

**Comment by julio-34727** [comment]
Hi @alexander-beedie,
parameter `engine_options` (header_row, skip_rows and so..;) must be passed in load_sheet_by_name/load_sheet_by_idx and not in read_excel (see fastexcel docs: https://fastexcel.toucantoco.dev/fastexcel.html)

**Comment by alexander-beedie** [comment]
> Hi @alexander-beedie, parameter `engine_options` (header_row, skip_rows and so..;) must be passed in load_sheet_by_name/load_sheet_by_idx and not in read_excel (see fastexcel docs: https://fastexcel.toucantoco.dev/fastexcel.html)

Thanks, will adjust accordingly; hadn't really gotten around to exposing those yet, as I'm planning to unify common options in the top-level function signature 👌

**Comment by adrivn** [comment]
@alexander-beedie Outstanding work. Those of us that *must* deal with Excel data cannot thank you enough for this addition.

**Comment by alexander-beedie** [comment]
> @alexander-beedie Outstanding work. Those of us that *must* deal with Excel data cannot thank you enough for this addition.

All the thanks here go to the `fastexcel` folks. I am merely the plumber on this one ;)

**Comment by SaelKimberly** [comment]
This is a great news for me, thanks! Excel files are my headache, really)

But, still, I have three main problems, and, I think, they are important and have simple solution - at the reader level. In my practice, these problems appeared frequently, and I think, not for me only.

https://github.com/pola-rs/polars/issues/14036

Thank you very much, polars is getting better and better, day by day 😌

**Comment by durgeksh** [comment]
I cannot thank you guys enough for polars. At my place almost everyone started using this as we saw an incredible performance over Pandas. 
I tried the new calamine engine, but found one issue with it. Although it is a new feature, thought to report.
https://github.com/pola-rs/polars/issues/14388

**Comment by alexander-beedie** [comment]
> I cannot thank you guys enough for polars. At my place almost everyone started using this as we saw an incredible performance over Pandas. I tried the new calamine engine, but found one issue with it. Although it is a new feature, thought to report. #14388

Thanks! Looking at your error report you should try and get a copy of your workbook to the `fastexcel` folks, so they can take a look and come up with a fix - we won't be able to fix it on our end (and it will be tricky without a sample workbook to test with - doesn't have tube your original data, just as long as it can reproduce the error) 👌

**Comment by durgeksh** [comment]
Thank you for this wonderful library. Sure, I will report to fastexcel.

**Comment by wuxianggujun** [comment]
I found that fastexcel does not support the merge cell function, which makes me very distressed. I have switched many engines

**Pull request merged**